dotCMS/src/main/java/com/dotmarketing/portlets/languagesmanager/business/LanguageFactoryImpl.java

```diff
@@ -152,27 +160,132 @@ protected Language getLanguage(final String languageId) {
-			language.setLanguageCode(Config.getStringProperty("DEFAULT_LANGUAGE_CODE", "en"));
+					Config.getStringProperty(DEFAULT_LANGUAGE_COUNTRY_CODE, "US"));
+			language.setLanguage(Config.getStringProperty(DEFAULT_LANGUAGE_STR, "English"));
+			language.setLanguageCode(Config.getStringProperty(DEFAULT_LANGUAGE_CODE, "en"));
 
 			//saves the new language
 			saveLanguage(language);
+
+			try {
+				writeDefaultLanguage(language.getId());
+				CacheLocator.getLanguageCache().setDefaultLanguage(language);
+			} catch (DotDataException e) {
+				throw new DotRuntimeException(String.format("Unable to persist language with id `%s` as the default in the default company.",language), e);
+			}
 		}
 
 
 		return language;
 
 	}
 
+	@Override
+	@WrapInTransaction
+	protected Language getDefaultLanguage() {
+
+		final LanguageCache languageCache = CacheLocator.getLanguageCache();
+		Language defaultLanguage = languageCache.getDefaultLanguage();
+
+		if(!LANG_404.equals(defaultLanguage)){
+			return defaultLanguage;
+		}
+
+		if (createDefaultLanguageLock.compareAndSet(null, defaultLanguage)) {
+			final long defaultLangId = readDefaultLanguage();
+			if (0 != defaultLangId) {
+				//Verify if it already exist. if so out put it into cache.
+				defaultLanguage = getLanguage(defaultLangId);
+			}
+			if (null == defaultLanguage) {
+				//if it doesn't creates the default lang and put it in cache.
+				defaultLanguage = createDefaultLanguage();
+			}
+			languageCache.setDefaultLanguage(defaultLanguage);
+			//update the lock.
+			createDefaultLanguageLock.set(defaultLanguage);
+		}
+
+		return createDefaultLanguageLock.get();
+	}
+
+
+	void writeDefaultLanguage(final long languageId) throws DotDataException {
+		final String companyId = APILocator.getCompanyAPI().getDefaultCompany().getCompanyId();
+		new DotConnect()
+				.setSQL("UPDATE company SET default_language_id = ? WHERE companyid = ? ")
+				.addParam(languageId)
+				.addParam(companyId)
+				.loadResult();
+	}
+
+	long readDefaultLanguage() {
+
+		final String companyId = APILocator.getCompanyAPI().getDefaultCompany().getCompanyId();
+		final String defaultLanguageId = new DotConnect()
+				.setSQL("SELECT default_language_id FROM company WHERE companyid = ? ")
+				.addParam(companyId).getString("default_language_id");
+		return ConversionUtils.toLong(defaultLanguageId);
+	}
+
+
+	@Override
+	@WrapInTransaction
+	protected void makeDefault(final Long languageId) throws DotDataException {
+		final Language language = getLanguage(languageId);
+		if(null == language ){
+		   throw new DotDataException(String.format("Unable to make default language with id %s the default one.",languageId));
+		}
+		if (createDefaultLanguageLock.compareAndSet(getDefaultLanguage(), language)) {
+			writeDefaultLanguage(languageId);
+			CacheLocator.getLanguageCache().clearCache();
+		}
+	}
+
+	@Override
+	@WrapInTransaction
+	protected void transferAssets(final Long oldDefaultLanguage, final Long newDefaultLanguage)
```

Comment:
not sure if the WrapInTranx should be here, rather than the API

Reply:
ok. I'll  change that